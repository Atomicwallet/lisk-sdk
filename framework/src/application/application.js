/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const assert = require('assert');
const {
	TransferTransaction,
	SecondSignatureTransaction,
	DelegateTransaction,
	VoteTransaction,
	MultisignatureTransaction,
	transactionInterface,
} = require('@liskhq/lisk-transactions');
const { getNetworkIdentifier } = require('@liskhq/lisk-cryptography');
const { validator: liskValidator } = require('@liskhq/lisk-validator');
const _ = require('lodash');
const Controller = require('../controller/controller');
const version = require('../version');
const validator = require('./validator');
const configurator = require('./default_configurator');
const { genesisBlockSchema, constantsSchema } = require('./schema');

const ApplicationState = require('./application_state');

const { createLoggerComponent } = require('../components/logger');
const { createStorageComponent } = require('../components/storage');
const { MigrationEntity } = require('../application/storage/entities');

const { Network } = require('./network');
const { NetworkInfoEntity } = require('../application/storage/entities');
const { networkMigrations } = require('../application/storage/migrations');

const Node = require('./node');

const { InMemoryChannel } = require('../controller/channels');

const HttpAPIModule = require('../modules/http_api');

const registerProcessHooks = app => {
	process.title = `${app.config.app.label}(${app.config.app.version})`;

	process.on('uncaughtException', err => {
		// Handle error safely
		app.logger.error(
			{
				err,
			},
			'System error: uncaughtException',
		);
		app.shutdown(1, err.message);
	});

	process.on('unhandledRejection', err => {
		// Handle error safely
		app.logger.fatal(
			{
				err,
			},
			'System error: unhandledRejection',
		);
		app.shutdown(1, err.message);
	});

	process.once('SIGTERM', () => app.shutdown(1));

	process.once('SIGINT', () => app.shutdown(1));

	process.once('cleanup', (error, code) => app.shutdown(code, error));

	process.once('exit', (error, code) => app.shutdown(code, error));
};

class Application {
	constructor(genesisBlock, config = {}) {
		const errors = liskValidator.validate(genesisBlockSchema, genesisBlock);
		if (errors.length) {
			throw errors;
		}

		// Don't change the object parameters provided
		let appConfig = _.cloneDeep(config);

		if (!_.has(appConfig, 'app.label')) {
			_.set(
				appConfig,
				'app.label',
				`lisk-${genesisBlock.payloadHash.slice(0, 7)}`,
			);
		}

		if (!_.has(appConfig, 'components.logger.logFileName')) {
			_.set(
				appConfig,
				'components.logger.logFileName',
				`${process.cwd()}/logs/${appConfig.app.label}/lisk.log`,
			);
		}

		appConfig = configurator.getConfig(appConfig, {
			failOnInvalidArg: process.env.NODE_ENV !== 'test',
		});

		// These constants are readonly we are loading up their default values
		// In additional validating those values so any wrongly changed value
		// by us can be catch on application startup
		const constants = validator.parseEnvArgAndValidate(constantsSchema, {});

		// app.genesisConfig are actually old constants
		// we are merging these here to refactor the underlying code in other iteration
		this.constants = { ...constants, ...appConfig.app.genesisConfig };
		this.genesisBlock = genesisBlock;
		this.config = appConfig;
		this.channel = null;
		this.initialState = null;
		this.applicationState = null;

		// TODO: This should be removed after https://github.com/LiskHQ/lisk/pull/2980
		global.constants = this.constants;
		this.logger = this._initLogger();
		this.storage = this._initStorage();

		// Private members
		this._modules = {};
		this._transactions = {};
		this._migrations = {};
		this._node = null;
		this._network = null;
		this._controller = null;

		this.registerTransaction(TransferTransaction);
		this.registerTransaction(SecondSignatureTransaction);
		this.registerTransaction(DelegateTransaction);
		this.registerTransaction(VoteTransaction);
		this.registerTransaction(MultisignatureTransaction);

		this.registerModule(Node, {
			registeredTransactions: this.getTransactions(),
		});
		this.registerModule(HttpAPIModule);
		this.overrideModuleOptions(HttpAPIModule.alias, {
			loadAsChildProcess: true,
		});
	}

	registerModule(moduleKlass, options = {}, alias = undefined) {
		assert(moduleKlass, 'ModuleSpec is required');
		assert(
			typeof options === 'object',
			'Module options must be provided or set to empty object.',
		);
		assert(alias || moduleKlass.alias, 'Module alias must be provided.');
		const moduleAlias = alias || moduleKlass.alias;
		assert(
			!Object.keys(this.getModules()).includes(moduleAlias),
			`A module with alias "${moduleAlias}" already registered.`,
		);

		this.config.modules[moduleAlias] = Object.assign(
			this.config.modules[moduleAlias] || {},
			options,
		);
		this._modules[moduleAlias] = moduleKlass;

		// Register migrations defined by the module
		this.registerMigrations(moduleKlass.alias, moduleKlass.migrations);
	}

	overrideModuleOptions(alias, options) {
		const modules = this.getModules();
		assert(
			Object.keys(modules).includes(alias),
			`No module ${alias} is registered`,
		);
		this.config.modules[alias] = {
			...this.config.modules[alias],
			...options,
		};
	}

	registerTransaction(Transaction, { matcher } = {}) {
		assert(Transaction, 'Transaction implementation is required');

		assert(
			Number.isInteger(Transaction.TYPE),
			'Transaction type is required as an integer',
		);

		assert(
			!Object.keys(this.getTransactions()).includes(
				Transaction.TYPE.toString(),
			),
			`A transaction type "${Transaction.TYPE}" is already registered.`,
		);

		validator.validate(transactionInterface, Transaction.prototype);

		if (matcher) {
			Object.defineProperty(Transaction.prototype, 'matcher', {
				get: () => matcher,
			});
		}

		this._transactions[Transaction.TYPE] = Object.freeze(Transaction);
	}

	registerMigrations(namespace, migrations) {
		assert(namespace, 'Namespace is required');
		assert(Array.isArray(migrations), 'Migrations list should be an array');
		assert(
			!Object.keys(this.getMigrations()).includes(namespace),
			`Migrations for "${namespace}" was already registered.`,
		);

		this._migrations[namespace] = Object.freeze(migrations);
	}

	getTransactions() {
		return this._transactions;
	}

	getTransaction(transactionType) {
		return this._transactions[transactionType];
	}

	getModule(alias) {
		return this._modules.get[alias];
	}

	getModules() {
		return this._modules;
	}

	getMigrations() {
		return this._migrations;
	}

	async run() {
		this.logger.info(
			'If you experience any type of error, please open an issue on Lisk GitHub: https://github.com/LiskHQ/lisk-sdk/issues',
		);
		this.logger.info(
			'Contribution guidelines can be found at Lisk-docs: https://github.com/LiskHQ/lisk-docs/blob/build/CONTRIBUTING.adoc',
		);
		this.logger.info(`Booting the application with Lisk Framework(${version})`);

		// Freeze every module and configuration so it would not interrupt the app execution
		this._compileAndValidateConfigurations();

		Object.freeze(this.genesisBlock);
		Object.freeze(this.constants);
		Object.freeze(this.config);

		this.logger.info(`Starting the app - ${this.config.app.label}`);

		registerProcessHooks(this);

		// Initialize all objects
		this.applicationState = this._initApplicationState();
		this.channel = this._initChannel();
		this.applicationState.channel = this.channel;

		this._controller = this._initController();
		this._network = this._initNetwork();

		// Load system components
		await this.storage.bootstrap();
		await this.storage.entities.Migration.defineSchema();

		await this._controller.load(
			this.getModules(),
			this.config.modules,
			this.getMigrations(),
		);

		await this._network.initialiseNetwork();

		this.channel.publish('app:ready');
	}

	async shutdown(errorCode = 0, message = '') {
		if (this._controller) {
			await this._controller.cleanup(errorCode, message);
		}
		this.logger.info({ errorCode, message }, 'Shutting down application');
		process.exit(errorCode);
	}

	// --------------------------------------
	// Private
	// --------------------------------------
	_compileAndValidateConfigurations() {
		const modules = this.getModules();
		this.config.app.networkId = getNetworkIdentifier(
			this.genesisBlock.payloadHash,
			this.genesisBlock.communityIdentifier,
		);

		const appConfigToShareWithModules = {
			version: this.config.app.version,
			minVersion: this.config.app.minVersion,
			protocolVersion: this.config.app.protocolVersion,
			networkId: this.config.app.networkId,
			genesisBlock: this.genesisBlock,
			constants: this.constants,
			lastCommitId: this.config.app.lastCommitId,
			buildVersion: this.config.app.buildVersion,
		};

		// TODO: move this configuration to module especific config file
		const childProcessModules = process.env.LISK_CHILD_PROCESS_MODULES
			? process.env.LISK_CHILD_PROCESS_MODULES.split(',')
			: [];

		Object.keys(modules).forEach(alias => {
			this.overrideModuleOptions(alias, {
				loadAsChildProcess: childProcessModules.includes(alias),
			});
			this.overrideModuleOptions(alias, appConfigToShareWithModules);
		});

		this.initialState = {
			version: this.config.app.version,
			minVersion: this.config.app.minVersion,
			protocolVersion: this.config.app.protocolVersion,
			networkId: this.config.app.networkId,
			wsPort: this.config.app.network.wsPort,
			httpPort: this.config.modules.http_api.httpPort,
		};

		this.logger.trace(this.config, 'Compiled configurations');
	}

	_initLogger() {
		return createLoggerComponent({
			...this.config.components.logger,
			module: 'lisk:app',
		});
	}

	_initStorage() {
		const storageConfig = this.config.components.storage;
		const loggerConfig = this.config.components.logger;
		const dbLogger =
			storageConfig.logFileName &&
			storageConfig.logFileName === loggerConfig.logFileName
				? this.logger
				: createLoggerComponent({
						...loggerConfig,
						logFileName: storageConfig.logFileName,
						module: 'lisk:app:database',
				  });

		const storage = createStorageComponent(
			this.config.components.storage,
			dbLogger,
		);

		storage.registerEntity('Migration', MigrationEntity);

		return storage;
	}

	_initApplicationState() {
		return new ApplicationState({
			initialState: this.initialState,
			logger: this.logger,
		});
	}

	_initChannel() {
		return new InMemoryChannel(
			'app',
			['ready', 'state:updated'],
			{
				getComponentConfig: {
					handler: action => this.config.components[action.params],
				},
				getApplicationState: {
					handler: () => this.applicationState.state,
				},
				updateApplicationState: {
					handler: action => this.applicationState.update(action.params),
				},
				sendToNetwork: {
					handler: action => this._network.send(action.params),
				},
				broadcastToNetwork: {
					handler: action => this._network.broadcast(action.params),
				},
				requestFromNetwork: {
					handler: action => this._network.request(action.params),
				},
				requestFromPeer: {
					handler: action => this._network.requestFromPeer(action.params),
				},
				getConnectedPeers: {
					handler: action => this._network.getConnectedPeers(action.params),
				},
				getDisconnectedPeers: {
					handler: action => this._network.getDisconnectedPeers(action.params),
				},
				applyPenaltyOnPeer: {
					handler: action => this._network.applyPenalty(action.params),
				},
			},
			{ skipInternalEvents: true },
		);
	}

	_initController() {
		return new Controller({
			appLabel: this.config.app.label,
			config: {
				components: this.config.components,
				ipc: this.config.app.ipc,
				tempPath: this.config.app.tempPath,
			},
			logger: this.logger,
			storage: this.storage,
			channel: this.channel,
		});
	}

	_initNetwork() {
		const network = new Network({
			options: this.config.app.network,
			storage: this.storage,
			logger: this.logger,
			channel: this.channel,
		});

		this.storage.registerEntity('NetworkInfo', NetworkInfoEntity);
		this.registerMigrations('network', networkMigrations());

		return network;
	}
}

module.exports = Application;
