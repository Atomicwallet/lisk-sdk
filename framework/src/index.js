const transactions = require('@liskhq/lisk-transactions');
const cryptography = require('@liskhq/lisk-cryptography');
const Application = require('./controller/application');
const BigNumber = require('./modules/chain/helpers/bignum');
const samples = require('../samples');
const version = require('./version');
const validator = require('./controller/helpers/validator');

/**
 * @namespace framework
 * @type {{constants, Application: (module.Application|*), version: string}}
 */
module.exports = {
	Application,
	BigNumber,
	version,
	...samples,
	helpers: {
		validator,
	},
	...transactions,
	cryptography,
};
