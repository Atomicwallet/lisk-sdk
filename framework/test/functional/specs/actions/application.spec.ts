/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { blockSchema, baseAccountSchema } from '@liskhq/lisk-chain';

import {
	TransferTransaction,
	DelegateTransaction,
	VoteTransaction,
	UnlockTransaction,
	MultisignatureTransaction,
	ProofOfMisbehaviorTransaction,
	BaseTransaction,
} from '@liskhq/lisk-transactions';
import { accountAssetSchema } from '../../../../src/application/node/account';

import { BlockProcessorV0 } from '../../../../src/application/node/block_processor_v0';
import { BlockProcessorV2 } from '../../../../src/application/node/block_processor_v2';

import { createApplication, closeApplication } from '../../utils/application';

import { Application } from '../../../../src';

describe('Application related actions', () => {
	let app: Application;

	beforeAll(async () => {
		app = await createApplication('actions-transactions');
	});

	afterAll(async () => {
		await closeApplication(app);
	});

	describe('getSchema', () => {
		it('should return schemas used to encode objects in framework', async () => {
			const frameworkSchemas = await app['_channel'].invoke('app:getSchema');

			const accountSchema = {
				...baseAccountSchema,
				properties: {
					...baseAccountSchema.properties,
					asset: {
						...baseAccountSchema.properties.asset,
						properties: accountAssetSchema,
					},
				},
			};

			const expectedFrameworkSchemas = {
				account: accountSchema,
				blockHeader: blockSchema,
				blockHeadersAssets: {
					0: BlockProcessorV0.schema,
					2: BlockProcessorV2.schema,
				},
				baseTransaction: BaseTransaction.BASE_SCHEMA,
				transactionsAssets: {
					[TransferTransaction.TYPE]: TransferTransaction.ASSET_SCHEMA,
					[DelegateTransaction.TYPE]: DelegateTransaction.ASSET_SCHEMA,
					[VoteTransaction.TYPE]: VoteTransaction.ASSET_SCHEMA,
					[UnlockTransaction.TYPE]: UnlockTransaction.ASSET_SCHEMA,
					[MultisignatureTransaction.TYPE]:
						MultisignatureTransaction.ASSET_SCHEMA,
					[ProofOfMisbehaviorTransaction.TYPE]:
						ProofOfMisbehaviorTransaction.ASSET_SCHEMA,
				},
			};

			expect(frameworkSchemas).toEqual(expectedFrameworkSchemas);
		});
	});
});
