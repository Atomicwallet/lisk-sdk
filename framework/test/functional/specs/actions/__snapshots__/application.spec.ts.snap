// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application related actions getNodeInfo should return node status and constants 1`] = `
Object {
  "finalizedHeight": 0,
  "genesisConfig": Object {
    "activeDelegates": 101,
    "baseFees": Array [
      Object {
        "assetType": 0,
        "baseFee": "1000000000",
        "moduleType": 5,
      },
    ],
    "bftThreshold": 68,
    "blockTime": 10,
    "communityIdentifier": "Lisk",
    "delegateListRoundOffset": 2,
    "maxPayloadLength": 15360,
    "minFeePerByte": 1000,
    "minRemainingBalance": "5000000",
    "rewards": Object {
      "distance": 3000000,
      "milestones": Array [
        "500000000",
        "400000000",
        "300000000",
        "200000000",
        "100000000",
      ],
      "offset": 2160,
    },
    "standbyDelegates": 2,
  },
  "height": 2,
  "lastBlockID": "jIl/x6qu3V7AfgULoWPeAHQA94wmuPqMe/8JhsAU1Sg=",
  "networkID": "hz2oWizucNpjHZCw8X+tqMOsm4OyYT9Mpf3dN00QNLM=",
  "networkVersion": "1.1",
  "syncing": false,
  "unconfirmedTransactions": 0,
  "version": "2.1.0",
}
`;

exports[`Application related actions getSchema should return schemas used to encode objects in framework 1`] = `
Object {
  "accountSchema": Object {
    "$id": "/account/base",
    "properties": Object {
      "address": Object {
        "dataType": "bytes",
        "fieldNumber": 1,
      },
      "dpos": Object {
        "fieldNumber": 5,
        "properties": Object {
          "delegate": Object {
            "fieldNumber": 1,
            "properties": Object {
              "consecutiveMissedBlocks": Object {
                "dataType": "uint32",
                "fieldNumber": 3,
              },
              "isBanned": Object {
                "dataType": "boolean",
                "fieldNumber": 5,
              },
              "lastForgedHeight": Object {
                "dataType": "uint32",
                "fieldNumber": 4,
              },
              "pomHeights": Object {
                "fieldNumber": 2,
                "items": Object {
                  "dataType": "uint32",
                },
                "type": "array",
              },
              "totalVotesReceived": Object {
                "dataType": "uint64",
                "fieldNumber": 6,
              },
              "username": Object {
                "dataType": "string",
                "fieldNumber": 1,
              },
            },
            "required": Array [
              "username",
              "pomHeights",
              "consecutiveMissedBlocks",
              "lastForgedHeight",
              "isBanned",
              "totalVotesReceived",
            ],
            "type": "object",
          },
          "sentVotes": Object {
            "fieldNumber": 2,
            "items": Object {
              "properties": Object {
                "amount": Object {
                  "dataType": "uint64",
                  "fieldNumber": 2,
                },
                "delegateAddress": Object {
                  "dataType": "bytes",
                  "fieldNumber": 1,
                },
              },
              "required": Array [
                "delegateAddress",
                "amount",
              ],
              "type": "object",
            },
            "type": "array",
          },
          "unlocking": Object {
            "fieldNumber": 3,
            "items": Object {
              "properties": Object {
                "amount": Object {
                  "dataType": "uint64",
                  "fieldNumber": 2,
                },
                "delegateAddress": Object {
                  "dataType": "bytes",
                  "fieldNumber": 1,
                },
                "unvoteHeight": Object {
                  "dataType": "uint32",
                  "fieldNumber": 3,
                },
              },
              "required": Array [
                "delegateAddress",
                "amount",
                "unvoteHeight",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "keys": Object {
        "fieldNumber": 4,
        "properties": Object {
          "mandatoryKeys": Object {
            "fieldNumber": 2,
            "items": Object {
              "dataType": "bytes",
            },
            "type": "array",
          },
          "numberOfSignatures": Object {
            "dataType": "uint32",
            "fieldNumber": 1,
          },
          "optionalKeys": Object {
            "fieldNumber": 3,
            "items": Object {
              "dataType": "bytes",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "sequence": Object {
        "fieldNumber": 3,
        "properties": Object {
          "nonce": Object {
            "dataType": "uint64",
            "fieldNumber": 1,
          },
        },
        "type": "object",
      },
      "token": Object {
        "fieldNumber": 2,
        "properties": Object {
          "balance": Object {
            "dataType": "uint64",
            "fieldNumber": 1,
          },
        },
        "type": "object",
      },
    },
    "required": Array [
      "address",
      "token",
      "sequence",
      "keys",
      "dpos",
    ],
    "type": "object",
  },
  "blockHeaderSchema": Object {
    "$id": "/block/header",
    "properties": Object {
      "asset": Object {
        "dataType": "bytes",
        "fieldNumber": 8,
      },
      "generatorPublicKey": Object {
        "dataType": "bytes",
        "fieldNumber": 6,
      },
      "height": Object {
        "dataType": "uint32",
        "fieldNumber": 3,
      },
      "previousBlockID": Object {
        "dataType": "bytes",
        "fieldNumber": 4,
      },
      "reward": Object {
        "dataType": "uint64",
        "fieldNumber": 7,
      },
      "signature": Object {
        "dataType": "bytes",
        "fieldNumber": 9,
      },
      "timestamp": Object {
        "dataType": "uint32",
        "fieldNumber": 2,
      },
      "transactionRoot": Object {
        "dataType": "bytes",
        "fieldNumber": 5,
      },
      "version": Object {
        "dataType": "uint32",
        "fieldNumber": 1,
      },
    },
    "required": Array [
      "version",
      "timestamp",
      "height",
      "previousBlockID",
      "transactionRoot",
      "generatorPublicKey",
      "reward",
      "asset",
    ],
    "type": "object",
  },
  "blockHeadersAssets": Object {
    "0": Object {
      "$id": "/genesisBlock/header/asset",
      "properties": Object {
        "accounts": Object {
          "fieldNumber": 1,
          "items": Object {
            "$id": "/account/base",
            "properties": Object {
              "address": Object {
                "dataType": "bytes",
                "fieldNumber": 1,
              },
              "dpos": Object {
                "fieldNumber": 5,
                "properties": Object {
                  "delegate": Object {
                    "fieldNumber": 1,
                    "properties": Object {
                      "consecutiveMissedBlocks": Object {
                        "dataType": "uint32",
                        "fieldNumber": 3,
                      },
                      "isBanned": Object {
                        "dataType": "boolean",
                        "fieldNumber": 5,
                      },
                      "lastForgedHeight": Object {
                        "dataType": "uint32",
                        "fieldNumber": 4,
                      },
                      "pomHeights": Object {
                        "fieldNumber": 2,
                        "items": Object {
                          "dataType": "uint32",
                        },
                        "type": "array",
                      },
                      "totalVotesReceived": Object {
                        "dataType": "uint64",
                        "fieldNumber": 6,
                      },
                      "username": Object {
                        "dataType": "string",
                        "fieldNumber": 1,
                      },
                    },
                    "required": Array [
                      "username",
                      "pomHeights",
                      "consecutiveMissedBlocks",
                      "lastForgedHeight",
                      "isBanned",
                      "totalVotesReceived",
                    ],
                    "type": "object",
                  },
                  "sentVotes": Object {
                    "fieldNumber": 2,
                    "items": Object {
                      "properties": Object {
                        "amount": Object {
                          "dataType": "uint64",
                          "fieldNumber": 2,
                        },
                        "delegateAddress": Object {
                          "dataType": "bytes",
                          "fieldNumber": 1,
                        },
                      },
                      "required": Array [
                        "delegateAddress",
                        "amount",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "unlocking": Object {
                    "fieldNumber": 3,
                    "items": Object {
                      "properties": Object {
                        "amount": Object {
                          "dataType": "uint64",
                          "fieldNumber": 2,
                        },
                        "delegateAddress": Object {
                          "dataType": "bytes",
                          "fieldNumber": 1,
                        },
                        "unvoteHeight": Object {
                          "dataType": "uint32",
                          "fieldNumber": 3,
                        },
                      },
                      "required": Array [
                        "delegateAddress",
                        "amount",
                        "unvoteHeight",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                },
                "type": "object",
              },
              "keys": Object {
                "fieldNumber": 4,
                "properties": Object {
                  "mandatoryKeys": Object {
                    "fieldNumber": 2,
                    "items": Object {
                      "dataType": "bytes",
                    },
                    "type": "array",
                  },
                  "numberOfSignatures": Object {
                    "dataType": "uint32",
                    "fieldNumber": 1,
                  },
                  "optionalKeys": Object {
                    "fieldNumber": 3,
                    "items": Object {
                      "dataType": "bytes",
                    },
                    "type": "array",
                  },
                },
                "type": "object",
              },
              "sequence": Object {
                "fieldNumber": 3,
                "properties": Object {
                  "nonce": Object {
                    "dataType": "uint64",
                    "fieldNumber": 1,
                  },
                },
                "type": "object",
              },
              "token": Object {
                "fieldNumber": 2,
                "properties": Object {
                  "balance": Object {
                    "dataType": "uint64",
                    "fieldNumber": 1,
                  },
                },
                "type": "object",
              },
            },
            "required": Array [
              "address",
              "token",
              "sequence",
              "keys",
              "dpos",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "initDelegates": Object {
          "fieldNumber": 2,
          "items": Object {
            "dataType": "bytes",
          },
          "minItems": 1,
          "type": "array",
        },
        "initRounds": Object {
          "dataType": "uint32",
          "fieldNumber": 3,
          "minimum": 3,
        },
      },
      "required": Array [
        "accounts",
        "initDelegates",
        "initRounds",
      ],
      "type": "object",
    },
    "2": Object {
      "$id": "/blockHeader/asset/v2",
      "properties": Object {
        "maxHeightPreviouslyForged": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
        "maxHeightPrevoted": Object {
          "dataType": "uint32",
          "fieldNumber": 2,
        },
        "seedReveal": Object {
          "dataType": "bytes",
          "fieldNumber": 3,
        },
      },
      "required": Array [
        "maxHeightPreviouslyForged",
        "maxHeightPrevoted",
        "seedReveal",
      ],
      "type": "object",
    },
  },
  "blockSchema": Object {
    "$id": "/block",
    "properties": Object {
      "header": Object {
        "dataType": "bytes",
        "fieldNumber": 1,
      },
      "payload": Object {
        "fieldNumber": 2,
        "items": Object {
          "dataType": "bytes",
        },
        "type": "array",
      },
    },
    "required": Array [
      "header",
      "payload",
    ],
    "type": "object",
  },
  "transactionSchema": Object {
    "$id": "lisk/transaction",
    "properties": Object {
      "asset": Object {
        "dataType": "bytes",
        "fieldNumber": 6,
      },
      "assetType": Object {
        "dataType": "uint32",
        "fieldNumber": 2,
      },
      "fee": Object {
        "dataType": "uint64",
        "fieldNumber": 4,
      },
      "moduleType": Object {
        "dataType": "uint32",
        "fieldNumber": 1,
        "minimum": 2,
      },
      "nonce": Object {
        "dataType": "uint64",
        "fieldNumber": 3,
      },
      "senderPublicKey": Object {
        "dataType": "bytes",
        "fieldNumber": 5,
      },
      "signatures": Object {
        "fieldNumber": 7,
        "items": Object {
          "dataType": "bytes",
        },
        "type": "array",
      },
    },
    "required": Array [
      "moduleType",
      "assetType",
      "nonce",
      "fee",
      "senderPublicKey",
      "asset",
    ],
    "type": "object",
  },
  "transactionsAssetSchemas": Array [
    Object {
      "assetType": 0,
      "moduleType": 2,
      "schema": Object {
        "$id": "lisk/transfer-asset",
        "properties": Object {
          "amount": Object {
            "dataType": "uint64",
            "fieldNumber": 1,
          },
          "data": Object {
            "dataType": "string",
            "fieldNumber": 3,
            "maxLength": 64,
            "minLength": 0,
          },
          "recipientAddress": Object {
            "dataType": "bytes",
            "fieldNumber": 2,
            "maxLength": 20,
            "minLength": 20,
          },
        },
        "required": Array [
          "amount",
          "recipientAddress",
          "data",
        ],
        "title": "Transfer transaction asset",
        "type": "object",
      },
    },
    Object {
      "assetType": 0,
      "moduleType": 4,
      "schema": Object {
        "$id": "lisk/keys/register",
        "properties": Object {
          "mandatoryKeys": Object {
            "fieldNumber": 2,
            "items": Object {
              "dataType": "bytes",
              "maxLength": 32,
              "minLength": 32,
            },
            "maxItems": 64,
            "minItems": 0,
            "type": "array",
          },
          "numberOfSignatures": Object {
            "dataType": "uint32",
            "fieldNumber": 1,
            "maximum": 64,
            "minimum": 1,
          },
          "optionalKeys": Object {
            "fieldNumber": 3,
            "items": Object {
              "dataType": "bytes",
              "maxLength": 32,
              "minLength": 32,
            },
            "maxItems": 64,
            "minItems": 0,
            "type": "array",
          },
        },
        "required": Array [
          "numberOfSignatures",
          "optionalKeys",
          "mandatoryKeys",
        ],
        "type": "object",
      },
    },
    Object {
      "assetType": 0,
      "moduleType": 5,
      "schema": Object {
        "$id": "lisk/dpos/register",
        "properties": Object {
          "username": Object {
            "dataType": "string",
            "fieldNumber": 1,
            "maxLength": 20,
            "minLength": 1,
          },
        },
        "required": Array [
          "username",
        ],
        "type": "object",
      },
    },
    Object {
      "assetType": 1,
      "moduleType": 5,
      "schema": Object {
        "$id": "lisk/dpos/vote",
        "properties": Object {
          "votes": Object {
            "fieldNumber": 1,
            "items": Object {
              "properties": Object {
                "amount": Object {
                  "dataType": "sint64",
                  "fieldNumber": 2,
                },
                "delegateAddress": Object {
                  "dataType": "bytes",
                  "fieldNumber": 1,
                  "maxLength": 20,
                  "minLength": 20,
                },
              },
              "required": Array [
                "delegateAddress",
                "amount",
              ],
              "type": "object",
            },
            "maxItems": 20,
            "minItems": 1,
            "type": "array",
          },
        },
        "required": Array [
          "votes",
        ],
        "type": "object",
      },
    },
    Object {
      "assetType": 2,
      "moduleType": 5,
      "schema": Object {
        "$id": "lisk/dpos/unlock",
        "properties": Object {
          "unlockObjects": Object {
            "fieldNumber": 1,
            "items": Object {
              "properties": Object {
                "amount": Object {
                  "dataType": "uint64",
                  "fieldNumber": 2,
                },
                "delegateAddress": Object {
                  "dataType": "bytes",
                  "fieldNumber": 1,
                  "maxLength": 20,
                  "minLength": 20,
                },
                "unvoteHeight": Object {
                  "dataType": "uint32",
                  "fieldNumber": 3,
                },
              },
              "required": Array [
                "delegateAddress",
                "amount",
                "unvoteHeight",
              ],
              "type": "object",
            },
            "maxItems": 20,
            "minItems": 1,
            "type": "array",
          },
        },
        "required": Array [
          "unlockObjects",
        ],
        "type": "object",
      },
    },
    Object {
      "assetType": 3,
      "moduleType": 5,
      "schema": Object {
        "$id": "lisk/dpos/pom",
        "properties": Object {
          "header1": Object {
            "$id": "lisk/block-header",
            "fieldNumber": 1,
            "properties": Object {
              "asset": Object {
                "fieldNumber": 8,
                "properties": Object {
                  "maxHeightPreviouslyForged": Object {
                    "dataType": "uint32",
                    "fieldNumber": 1,
                  },
                  "maxHeightPrevoted": Object {
                    "dataType": "uint32",
                    "fieldNumber": 2,
                  },
                  "seedReveal": Object {
                    "dataType": "bytes",
                    "fieldNumber": 3,
                  },
                },
                "required": Array [
                  "maxHeightPreviouslyForged",
                  "maxHeightPrevoted",
                  "seedReveal",
                ],
                "type": "object",
              },
              "generatorPublicKey": Object {
                "dataType": "bytes",
                "fieldNumber": 6,
              },
              "height": Object {
                "dataType": "uint32",
                "fieldNumber": 3,
              },
              "previousBlockID": Object {
                "dataType": "bytes",
                "fieldNumber": 4,
              },
              "reward": Object {
                "dataType": "uint64",
                "fieldNumber": 7,
              },
              "signature": Object {
                "dataType": "bytes",
                "fieldNumber": 9,
              },
              "timestamp": Object {
                "dataType": "uint32",
                "fieldNumber": 2,
              },
              "transactionRoot": Object {
                "dataType": "bytes",
                "fieldNumber": 5,
              },
              "version": Object {
                "dataType": "uint32",
                "fieldNumber": 1,
              },
            },
            "required": Array [
              "version",
              "timestamp",
              "height",
              "previousBlockID",
              "transactionRoot",
              "generatorPublicKey",
              "reward",
              "asset",
            ],
            "type": "object",
          },
          "header2": Object {
            "$id": "lisk/block-header",
            "fieldNumber": 2,
            "properties": Object {
              "asset": Object {
                "fieldNumber": 8,
                "properties": Object {
                  "maxHeightPreviouslyForged": Object {
                    "dataType": "uint32",
                    "fieldNumber": 1,
                  },
                  "maxHeightPrevoted": Object {
                    "dataType": "uint32",
                    "fieldNumber": 2,
                  },
                  "seedReveal": Object {
                    "dataType": "bytes",
                    "fieldNumber": 3,
                  },
                },
                "required": Array [
                  "maxHeightPreviouslyForged",
                  "maxHeightPrevoted",
                  "seedReveal",
                ],
                "type": "object",
              },
              "generatorPublicKey": Object {
                "dataType": "bytes",
                "fieldNumber": 6,
              },
              "height": Object {
                "dataType": "uint32",
                "fieldNumber": 3,
              },
              "previousBlockID": Object {
                "dataType": "bytes",
                "fieldNumber": 4,
              },
              "reward": Object {
                "dataType": "uint64",
                "fieldNumber": 7,
              },
              "signature": Object {
                "dataType": "bytes",
                "fieldNumber": 9,
              },
              "timestamp": Object {
                "dataType": "uint32",
                "fieldNumber": 2,
              },
              "transactionRoot": Object {
                "dataType": "bytes",
                "fieldNumber": 5,
              },
              "version": Object {
                "dataType": "uint32",
                "fieldNumber": 1,
              },
            },
            "required": Array [
              "version",
              "timestamp",
              "height",
              "previousBlockID",
              "transactionRoot",
              "generatorPublicKey",
              "reward",
              "asset",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "header1",
          "header2",
        ],
        "type": "object",
      },
    },
  ],
}
`;
